openml/extensions/extension_interface.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional, Tuple
+
+from openml import OpenMLTask, OpenMLFlow
+from openml.runs.trace import OpenMLTraceIteration, OpenMLRunTrace
+
+
+class Extension(ABC):
+
+    """Defines the interface to connect machine learning libraries to OpenML-Python.
+    """
+
+    ################################################################################################
+    # Abstract methods for flow serialization and de-serialization
+
+    @abstractmethod
+    def flow_to_model(self, flow: OpenMLFlow) -> Any:
+        pass
+
+    @abstractmethod
+    def model_to_flow(self, model: Any) -> OpenMLFlow:
+        pass
+
+    @abstractmethod
+    def flow_to_parameters(self, model: Any) -> List:
+        pass
+
+    @abstractmethod
+    def is_estimator(self, model: Any) -> bool:
+        pass
+
+    @abstractmethod
+    def seed_model(self, model: Any, seed: Optional[int]) -> Any:
+        pass
+
+    @abstractmethod
+    def run_model_on_fold(
+        self,
+        model: Any,
+        task: OpenMLTask,
+        rep_no: int,
+        fold_no: int,
+        sample_no: int,
+        can_measure_runtime: bool,
+        add_local_measures: bool,
+        extension: 'Extension',
+    ) -> Tuple:
+        pass
+
+    ################################################################################################
+    # Abstract methods for hyperparameter optimization
+
+    @abstractmethod
+    def is_hpo_class(self, model: Any) -> bool:
+        pass
+
+    @abstractmethod
+    def assert_hpo_class(self, model: Any) -> None:
+        if not self.is_hpo_class(model):
+            raise ValueError("Flow model is not a hyperparameter optimization algorithm.")
+
+    @abstractmethod
+    def assert_hpo_class_has_trace(self, model: Any) -> None:
+        pass
+
+    @abstractmethod
+    def instantiate_model_from_hpo_class(
+        self,
+        model: Any,
+        trace_iteration: OpenMLTraceIteration,
+    ) -> Any:
+        pass
+
+    @abstractmethod
+    def obtain_arff_trace(
```

Comment:
I wonder whether this method is really necessary. In fact, this method might as well be the same for all trace iterations, and make calls to `extract_trace_data` (`extract_trace_data` is of course necessary).

Reply:
I'm not sure if I get your comment on how this could be the same for all trace iterations. This method is actually used by `openml.runs.functions` and I guess it is pretty helpful to assemble the final trace. The method `extract_trace_data` on the other hand is not necessary in the abstract extension. Therefore, I removed it.

Reply:
> I assume we have the same idea about this, right?

Yes, I agree that we can put the merging of the traces into the general part of the library. I will add this to the proposed changes list for the updates to the extension interface.

Reply:
Where can I find this list?

Reply:
In an unsaved tab in gedit on my laptop ;)

Here it is:
* Remove `obtain_arff_trace` and return the trace headers in each fold and later on check that they're all the same.
* Simplify `run_model_on_fold` to accept X and y and do not split the data itself
* Simplify the return value of `run_model_on_fold` to have functionality for transforming the predictions in the actual task classes.
* Figure out how to best handle dependencies (as raised [here](https://github.com/openml/openml-python/pull/647#discussion_r272686627))